apps/sgx/enclave.cc

```diff
@@ -1,62 +1,32 @@
 #include <dlpack/dlpack.h>
+#include <tvm/runtime/c_runtime_api.h>
 #include <tvm/runtime/module.h>
 #include <tvm/runtime/registry.h>
 #include <tvm/runtime/packed_func.h>
-#ifndef _LIBCPP_SGX_CONFIG
-#include <iostream>
-#endif
-
-/* This function mirrors the one in howto_deploy except without the iostream */
-int Verify(tvm::runtime::Module mod, std::string fname) {
-  // Get the function from the module.
-  tvm::runtime::PackedFunc f = mod.GetFunction(fname);
-
-  // Allocate the DLPack data structures.
-  DLTensor* x;
-  DLTensor* y;
-  int ndim = 1;
-  int dtype_code = kDLFloat;
-  int dtype_bits = 32;
-  int dtype_lanes = 1;
-  int device_type = kDLCPU;
-  int device_id = 0;
-  int64_t shape[1] = {10};
-  TVMArrayAlloc(shape, ndim, dtype_code, dtype_bits, dtype_lanes,
-                device_type, device_id, &x);
-  TVMArrayAlloc(shape, ndim, dtype_code, dtype_bits, dtype_lanes,
-                device_type, device_id, &y);
-  for (int i = 0; i < shape[0]; ++i) {
-    static_cast<float*>(x->data)[i] = i;
-  }
-
-  // Invoke the function
-  f(x, y);
-
-  // check the output
-  bool all_eq = true;
-  for (int i = 0; i < shape[0]; ++i) {
-    all_eq = all_eq && static_cast<float*>(y->data)[i] == i + 1.0f;
-  }
-
-  return all_eq;
-}
+#include "../../src/runtime/sgx/trusted/runtime.h"
+#include "../../src/runtime/sgx/trusted/runtime.cc"
 
+using namespace tvm::runtime;
 
 extern "C" {
-void tvm_ecall_run_module(const void* tvm_args, void* tvm_return_value) {
-  tvm::runtime::Module mod_syslib = (*tvm::runtime::Registry::Get("module._GetSystemLib"))();
-  *(int*)tvm_return_value = Verify(mod_syslib, "addonesys");
-}
+void tvm_ecall_init() {}
+
+void tvm_ecall_packed_func(const char* cname,
```

Comment:
I understand the cost of ecall is high so it may not help at all. But it is still desirable to have a specific function index instead of look it up by name every-time we call. 

This might require a special global SGX registry that keeps index -> function, and call of function pays only a single table lookup

Reply:
actually we want something simpler than global singleton, as we only need to create a global table and report id back to the registry by change 
tvm_ocall_register_func to 
```c++
tvm_ocall_register_func(const char* name, int eid); 
```

Reply:
right, but that would require making a second global `Registry` within the enclave. I wasn't too keen on messing with all of that code, but it would have been the correct approach.